packages/jest-resolve/src/index.js

```diff
@@ -134,23 +136,29 @@ class Resolver {
     }
 
     // 0. If we have already resolved this module for this directory name,
-    //    return a value from the cache.
+    // return a value from the cache.
     if (this._moduleNameCache[key]) {
       return this._moduleNameCache[key];
     }
 
     // 1. Check if the module is a haste module.
-    let module = this.getModule(moduleName);
+    module = this.getModule(moduleName);
+    if (module) {
+      return (this._moduleNameCache[key] = module);
+    }
+
+    // 2. Check if the module is a haste module (but now, in the mocks).
+    module = this.getModuleMock(dirname + path.sep + 'index.js', moduleName);
```

Comment:
is this `index.js` safe here?

Reply:
It is. `getMockModule` uses it to pass it to `_resolveStubModuleName`, which does a `dirname` again. It's unfortunate that `getModule` and `getMockModule` have such a different interface; but I'm going to not refactor it for now.